### ISSUE TITLE
Semantic Segmentation Task Trainer 

### ISSUE THREAD

**Issue by Hamish-Cam** [microsoft/torchgeo#453]
Hello, firstly I apologise for the vagueness of the question - but having dug into this issue for a couple of days I haven't been able to understand it very well and so defining it as clearly as I would like has been difficult. I would like to know if there is some mistake in my training setup that is causing the issue outlined below or whether my understanding of the relationship between the sampler, dataloader and trainer is incorrect. 

Task background: I have a single channel input layer and a single channel mask output layer. I am using a randomBatchGeoSampler for getting training samples, which has been tested by printing/plotting samples obtained from the dataloader I am using. Currently I am using a `batch_size = 2` and `length = 2` (one batch per epoch) as a simple test case to see if I can get my trainer to run. In this case I receive 2 samples in total as expected.

I have then defined a trainer object as follows: 

```
model = SemanticSegmentationTask(
    segmentation_model="unet",
    encoder_name="resnet18",
    encoder_weights="imagenet",
    in_channels=1,
    num_classes=2,
    num_filters=32,
    loss="ce",
    ignore_zeros=False,
    learning_rate=0.1,
    learning_rate_schedule_patience=5,
)

wandb_logger = WandbLogger(project="Wildfires", log_model="all", name="simple_land_simple_dataloader")

trainer = Trainer(gpus=1, logger=wandb_logger,
                  callbacks=callbacks,
                  max_epochs=1,
                  precision=16,
                  log_every_n_steps=1,
                  max_steps=1,
               )

wandb_logger.watch(model)

trainer.fit(model=model, datamodule=datamodule)
```

Expected behaviour: I expect the trainer to run for a very short time, given that only one batch of 2 samples is being used and only 1 training step will be completed. I expect that the length of each training step should not increase with the size of the dataset used since the random samples taken are of equal patch and batch sizes. My current understanding is that the progress metric (6081025 in the picture below) should be the total batches (or something similar). 

Observed behaviour: The single training step results in a progress bar appearing where the progress metric is measured in the millions (meanwhile actual progress increases slowly in increments of 20), resulting in a massively long training process. Furthermore, the progress metric increases to larger and larger values as I attempt to add more data into my dataset. As far as I can see, the amount of data used in my dataset is the only thing that effects the size of this progress metric (including the segmentation model used). If I set max_steps = 0, the training loop does not run at all, as expected. 

<img width="1318" alt="Screenshot 2022-03-09 at 12 09 54" src="https://user-images.githubusercontent.com/91635473/157477467-43d21ee7-2b22-44f3-83e4-29efdfe93a49.png">

**Comment by Hamish-Cam** [comment]
Also happy to post the dataset classes if required as well.

In script use: 
```
batch_size=2
sampler_size =256
length = 2

datamodule = MODISJDLandcoverSimpleDataModule(
    modis_root_dir="MODIS/Test/",
    landcover_root_dir="landcover/",
    patch_size=sampler_size,
    length=length,
    batch_size=batch_size,
    num_workers=2,
    one_hot_encode=False,
    balance_samples=False,
    burn_prop = 0.5,
    units = Units.PIXELS
)
```

Datamodule code:
```
from typing import Any, Dict, Optional

from ..data_loading import LandcoverSimple, MODIS_JD, Sentinel2
from ..samplers import ConstrainedRandomBatchGeoSampler

import torch
import pytorch_lightning as pl
import torch.nn.functional as F
from torch.utils.data import DataLoader
from torchgeo.samplers.batch import RandomBatchGeoSampler
from torchgeo.samplers.single import GridGeoSampler
from torchgeo.datasets import stack_samples
from torchgeo.samplers.constants import Units


class MODISJDLandcoverSimpleDataModule(pl.LightningDataModule):

    # TODO: tune these hyperparams
    # stride is not trivial to choose
    # patch_size - stride should approx equal
    # the receptive field of a CNN

    stride = 5

    simple_classes = {
        "invalid": 0,
        "deciduous forests": 1,
        "coniferous forests": 2,
        "swamp forests": 3,
        "meadows": 4,
        "agriculture": 5,
        "bogs and mires": 6,
        "clearings and cuttings": 7,
        "water": 8,
        "urban": 9,
    }

    def __init__(
        self,
        modis_root_dir: str,
        landcover_root_dir: str,
        sentinel_root_dir: Optional[str] = None,
        batch_size: int = 64,
        length: int = 256,
        num_workers: int = 0,
        patch_size: int = 256,  # dav version
        # patch_size: int = 0.0459937425469195,  # stable version
        one_hot_encode: bool = False,
        balance_samples: bool = False,  # whether or not to constrain the sampler
        burn_prop: float = 0.5,
        units: Units = Units.PIXELS,
        **kwargs: Any,
    ) -> None:
        """Initialize a LightningDataModule for MODIS and Landcover based DataLoaders.
        Args:
            modis_root_dir: directory containing MODIS data
            landcover_root_dir: directory containing (Polesia) landcover data
            batch_size: number of samples in batch
            num_workers:
            patch_size:
            one_hot_encode: set True to one hot encode landcover classes
            balance_samples: set True to get more samples with fires in training
        """
        super().__init__()  # type: ignore[no-untyped-call]
        self.modis_root_dir = modis_root_dir
        self.landcover_root_dir = landcover_root_dir
        self.sentinel_root_dir = sentinel_root_dir

        self.batch_size = batch_size
        self.length = length
        self.num_workers = num_workers
        self.patch_size = patch_size
        self.one_hot_encode = one_hot_encode
        self.balance_samples = balance_samples
        self.burn_prop = burn_prop
        self.units = units

    def modis_transforms(self, sample: Dict[str, Any]) -> Dict[str, Any]:
        # Binarize the samples
        sample["mask"] = torch.where(
            sample["mask"] > 0,
            torch.ones(sample["mask"].shape, dtype=torch.int16),  # torchgeo dev req
            # torch.ones(sample["mask"].shape, dtype=torch.int32),
            sample["mask"],
        )
        sample["mask"] = torch.where(
            sample["mask"] < 0,
            torch.zeros(sample["mask"].shape, dtype=torch.int16),  # torchgeo dev req
            # torch.zeros(sample["mask"].shape, dtype=torch.int32),
            sample["mask"],
        )

        sample_ = {"mask": sample["mask"].long().squeeze()}

        return sample_

    def landcover_transforms(
        self,
        sample: Dict[str, Any],
    ) -> Dict[str, Any]:

        if self.one_hot_encode:

            x_shape = sample["image"].shape

            input_mask = sample["image"]
            # Get the one-hot encodings, which are a tensor of: (batch_idx, lat, lon, band_idx)
            encodings = F.one_hot(
                input_mask.to(torch.int64), num_classes=len(self.simple_classes)
            )

            # not sure this is the right shape we want for the mask – here the batch idx is missing
            # but maybe that gets added in the dataloader
            new_mask = torch.zeros([len(self.simple_classes), x_shape[1], x_shape[2]])
            for band_idx in range(0, len(self.simple_classes)):
                new_mask[band_idx, :, :] = encodings[0, :, :, band_idx]

            # Write one-hot encodings to the output tensor
            # outputs = sample
            sample["image"] = new_mask.to(torch.int32)

        sample_ = {"image": sample["image"].float()}

        return sample_

    def setup(self, stage: Optional[str] = None) -> None:
        """Initialize the main ``Dataset`` objects.
        This method is called once per GPU per run.
        Args:
            stage: state to set up
        """
        # consider whether we should pass the res and crs from one to another here

        landcover = LandcoverSimple(
            self.landcover_root_dir,
            transforms=self.landcover_transforms,
        )

        modis = MODIS_JD(
            self.modis_root_dir,
            landcover.crs,
            landcover.res,
            transforms=self.modis_transforms,
        )

        self.dataset = landcover & modis

        if self.sentinel_root_dir is not None:
            sentinel = Sentinel2(
                self.sentinel_root_dir,
                landcover.crs,
                landcover.res,
                bands=["B03", "B04", "B08"],
            )
            self.dataset = self.dataset & sentinel

        roi = self.dataset.bounds

        if self.balance_samples:
            print("Using a constrained sampler to get more samples with fires.")
            self.train_sampler = ConstrainedRandomBatchGeoSampler(
                dataset=self.dataset,
                size=self.patch_size,
                batch_size=self.batch_size,
                length=self.length,
                burn_prop=self.burn_prop,
                roi=roi,
                units=self.units,
            )

        else:
            self.train_sampler = RandomBatchGeoSampler(
                self.dataset, self.patch_size, self.batch_size, self.length, roi
            )

        self.val_sampler = GridGeoSampler(
            self.dataset, self.patch_size, self.stride, roi
        )
        self.test_sampler = GridGeoSampler(
            self.dataset, self.patch_size, self.stride, roi
        )

    def train_dataloader(self) -> DataLoader[Any]:
        """Return a DataLoader for training.
        Returns:
            training data loader
        """
        return DataLoader(
            self.dataset,
            batch_sampler=self.train_sampler,
            num_workers=self.num_workers,
            collate_fn=stack_samples,
        )

    def val_dataloader(self) -> DataLoader[Any]:
        """Return a DataLoader for validation.
        Returns:
            validation data loader
        """
        return DataLoader(
            self.dataset,
            batch_size=self.batch_size,
            sampler=self.val_sampler,
            num_workers=self.num_workers,
            collate_fn=stack_samples,
            shuffle=False,
        )

    def test_dataloader(self) -> DataLoader[Any]:
        """Return a DataLoader for testing.
        Returns:
            testing data loader
        """
        return DataLoader(
            self.dataset,
            batch_size=self.batch_size,
            sampler=self.test_sampler,
            num_workers=self.num_workers,
            collate_fn=stack_samples,
            shuffle=False,
        )

```

**Comment by isaaccorley** [comment]
@Hamish-Cam I'd recommend also posting this as an issue on PyTorch Lightning. We don't have any control over whether there's a bug in their trainer class (e.g. a potential bug caused if max_steps=1).

**Comment by calebrob6** [comment]
My first guess (haven't looked too closely) is that it is trying to spend all of this time in the validation step. What happens if you replace `self.val_sampler` and `self.test_sampler` with similar random batch geo samplers?

**Comment by Hamish-Cam** [comment]
@calebrob6 thanks for this - it does indeed seem that this was the case. I'm assuming this is due to the stride/patch_size parameter combination we were using. 

Have you any advice/guidance on how essential the grid sampler is for val/test and how best to pick these parameters accordingly?

**Comment by calebrob6** [comment]
If you have a _lot_ of validation data then I think it makes the most sense to sample from it (i.e. with the RandomBatchGeoSampler), else you should use large patch sizes to minimize edge effects.

**Comment by adamjstewart** [comment]
> Have you any advice/guidance on how essential the grid sampler is for val/test

Note that for the same number of patches, GridGeoSampler is significantly faster than RandomGeoSampler (see figure 3A: https://arxiv.org/abs/2111.08872) thanks to cache reuse. I agree that your stride is likely too small and that's causing your epoch length to be very large.

**Comment by adamjstewart** [comment]
@Hamish-Cam are you still having issues or did you figure this one out?

**Comment by Hamish-Cam** [comment]
All resolved! @calebrob6 suggestion did the trick
